nixos/invidious: cleanup, allow for easy scaling and load balancing, add http3-ytproxy

999eagle opened the pull request:
## Description of changes

This PR contains mainly three changes:
- The database maintenance service has been removed as it is obsolete, see https://docs.invidious.io/db-maintenance/.
- The new options `services.invidious.serviceScale` and `services.invidious.address` have been introduced to allow scaling Invidious to multiple services as recommended in https://docs.invidious.io/improve-public-instance/#2-multiple-invidious-processes and to allow configuring Invidious to not listen on `0.0.0.0`.
- The new option `services.invidious.http3-ytproxy.enable` has been introduced to configure `http3-ytproxy` according to https://docs.invidious.io/improve-public-instance/#3-speed-up-video-playback-with-http3-ytproxy.

Tests have also been fixed by switching to PostgreSQL 14 (see #216989) and new tests for the scaled Invidious deployment and http3-ytproxy have been added.

I've deployed these changes to my own Invidious instance at iv.catgirl.cloud for testing.

CC: @sbruder @GaetanLepage @infinisil 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

999eagle committed:
nixos/invidious: remove obsolete database maintenance service

See the note on https://docs.invidious.io/db-maintenance/. Manual maintenance
is no longer required.

999eagle committed:
nixos/invidious: add option to run more invidious instances

999eagle committed:
nixos/tests/invidious: move postgres-tcp into second machine and fix tests

Using PostgreSQL 15 without the init script fails due to
https://github.com/NixOS/nixpkgs/issues/216989.

999eagle committed:
nixos/tests/invidious: add test for scaled invidious

999eagle committed:
nixos/invidious: bind to 127.0.0.1 instead of 0.0.0.0 if nginx is used

999eagle committed:
http3-ytproxy: init at unstable-2022-07-03

999eagle committed:
nixos/invidious: add option to configure http3-ytproxy for invidious

SuperSandro2000 commented on pkgs/by-name/ht/http3-ytproxy/package.nix:
```diff
@@ -0,0 +1,49 @@
+      --replace 'lucas-clemente/quic-go' 'quic-go/quic-go'
+  '';
+
+  overrideModAttrs = _: {
+    preBuild = ''
+      substituteInPlace go.mod \
+        --replace 'go 1.16' 'go 1.20'
+      go mod tidy
+    '';
+    postInstall = ''
+      cp go.mod "$out/go.mod"
+    '';
+  };
```
Why do we need to bump the go version in go.mod? Normally this shouldn't be required. Also I think it would be easier if do this in postPatch which should also avoid the preBuild phase.

SuperSandro2000 commented on pkgs/by-name/ht/http3-ytproxy/package.nix:
```diff
@@ -0,0 +1,49 @@
+  };
+
+  postPatch = ''
+    substituteInPlace main.go \
+      --replace 'lucas-clemente/quic-go' 'quic-go/quic-go'
```
We should do this via a patch. Go is rather particular about it's files.

999eagle replied:
I've changed this dependency update, the go bump and running `go mod tidy` to be applied through a patch.

SuperSandro2000 commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -235,6 +304,16 @@ in
+    address = lib.mkOption {
+      type = types.str;
+      # default from https://github.com/iv-org/invidious/blob/master/config/config.example.yml
+      default = if cfg.nginx.enable then "127.0.0.1" else "0.0.0.0";
```
Either we do not open the firewall automatically, do we? Then this should be fine, I guess.

SuperSandro2000 commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -162,14 +203,28 @@ let
-        locations."/".proxyPass = "http://127.0.0.1:${toString cfg.port}";
+        locations."/".proxyPass =
+          if cfg.serviceScale == 1 then
+            "http://${ip}:${toString cfg.port}"
```
I don't think you can proxy to 0.0.0.0, unless this is something special in nginx but with normal routing in mind this should fail. Did you explicitly test this?

SuperSandro2000 commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -130,28 +167,32 @@ let
-        User = "invidious";
+    services.nginx.virtualHosts.${cfg.domain} = lib.mkIf cfg.nginx.enable {
+      locations."~ (^/videoplayback|^/vi/|^/ggpht/|^/sb/)" = {
+        proxyPass = "http://unix:/run/http3-ytproxy/socket/http-proxy.sock";
```
Nice that we are using unix sockets 👍🏼

SuperSandro2000 commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -94,14 +123,22 @@ let
+        assertion = cfg.database.host != null -> cfg.database.passwordFile != null;
+        message = "If database host isn't null, database password needs to be set";
+      }
+      {
+        assertion = cfg.serviceScale >= 1;
+        message = "Service can't be scaled below one instance";
+      }
```
I think we can avoid this by using a uint type for the option.

SuperSandro2000 replied:
🤔 maybe just set it to somethign cartoonishly large like 1024 but we are trading one downside for another. Either is fine I guess.

RaitoBezarius replied:
I'd argue that scaling down to zero is a feature but either case, this is completely enough and sufficient.

999eagle replied:
Well scaling down to zero is basically the same as setting `enable = false`. Except that we're still (optionally) provisioning a local PostgreSQL database and generating an invalid nginx configuration. I don't know what this would be useful for to be honest.

SuperSandro2000 commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -10,77 +10,106 @@ let
+      # randomized extra offset of up to 5 minutes is added.
+      Restart = lib.mkDefault "always";
+      RuntimeMaxSec = lib.mkDefault "1h";
+      RuntimeRandomizedExtraSec = lib.mkDefault "5min";
```
Very important to prevent dribble down effects!

SuperSandro2000 commented on nixos/tests/invidious.nix:
```diff
@@ -5,49 +5,75 @@ import ./make-test-python.nix ({ pkgs, ... }: {
+    postgres-tcp = { config, pkgs, ... }: {
+      services.postgresql = {
+        enable = true;
+        package = pkgs.postgresql_14;
```
Since the default in nixos is now 15, I don't feel that great to only downgrade it in the tests. Can we apply one of the workaround in the init script instead?

999eagle replied:
I've changed this now to run with the default PostgreSQL 15 and using an init script.

infinisil commented:
Generally looks good! I don't have the time for a detailed review, but ping me before it gets stuck too long :)

999eagle committed:
nixos/invidious: change default database user to invidious

This makes sure we don't need any workarounds for running Invidious with a local
PostgreSQL database.
Changing the default user should be fine as the new init script for PostgreSQL automatically
creates the new user and changes the existing database's owner to the new user. The old user
will still linger and must be removed manually.
See also: https://github.com/NixOS/nixpkgs/pull/266270

999eagle commented:
@RaitoBezarius I've changed this module to use `ensureDBOwnership = true` now as was requested in #266270.

@infinisil it'd be great if you could review this :)

RaitoBezarius commented:
> @RaitoBezarius I've changed this module to use `ensureDBOwnership = true` now as was requested in #266270.
> 
> @infinisil it'd be great if you could review this :)

Thank you so much!!

RaitoBezarius commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -114,7 +114,7 @@ let
 
       db = {
-        user = lib.mkDefault "kemal";
+        user = lib.mkDefault "invidious";
```
If you do it this way, BTW, you have to notice that you will break every existing setups that defaulted on kemal.

999eagle replied:
I thought about using `lib.versionAtLeast config.system.stateVersion "23.11"` for this to only default to `"invidious"` for newer setups but 1) the postgres module will just create the new user and change the owner of the existing db to the new user without breakage and 2) externally provisioned databases most likely configure this explicitly anyway. I've added a corresponding entry to the release notes for that. If you want, we can still switch this to check the state version?

RaitoBezarius replied:
I think it would be good to have the `stateVersion` style stuff if possible.

999eagle replied:
Done!

999eagle commented:
Updated so the database changes now only apply by default for `system.stateVersion >= 24.05`.

GaetanLepage commented:
> Updated so the database changes now only apply by default for `system.stateVersion >= 24.05`.

What if a user is following `nixos-unstable`.
He might still have `stateVersion = 23.11` in his config but at the same time he would be running a recent version of invidious.
Wouldn't it complain about not finding the database ?

999eagle commented:
If `stateVersion` is anything less than `24.05`, the default database user stays `kemal` to not break compatibility with externally provisioned databases. For automatically provisioned databases the assertion is triggered to prompt users to manually set the database user to `invidious` which will then create the new user in postgres and change the owner of the database to the new user. I think this should be enough to not break compatibility.

GaetanLepage commented:
> If `stateVersion` is anything less than `24.05`, the default database user stays `kemal` to not break compatibility with externally provisioned databases. For automatically provisioned databases the assertion is triggered to prompt users to manually set the database user to `invidious` which will then create the new user in postgres and change the owner of the database to the new user. I think this should be enough to not break compatibility.

Oh I see ! I missed the warning part. Thanks :)

SuperSandro2000 commented on nixos/tests/invidious.nix:
```diff
@@ -5,48 +5,72 @@ import ./make-test-python.nix ({ pkgs, ... }: {
+        enable = true;
+        initialScript = pkgs.writeText "init-postgres-with-password" ''
+          CREATE USER invidious WITH PASSWORD 'correct horse battery staple';
+          CREATE DATABASE invidious WITH OWNER invidious;
```
Not sure if this could use ensureDBOwnership

RaitoBezarius replied:
This is an... Explicit TCP/IP test?

999eagle replied:
I'm not really sure why we're testing PostgreSQL over TCP to be honest. I think the test was originally meant to test the password file which requires TCP instead of a unix socket.
I moved this to a separate VM so the test runs a bit faster as the secondary VM can start PostgreSQL while the first VM runs the different nginx tests.

SuperSandro2000 commented on nixos/modules/services/web-apps/invidious.nix:
```diff
@@ -10,82 +10,115 @@ let
+        ''
+        # configure threads for secondary instances
+        + lib.optionalString (scaleIndex > 0) ''
+          configParts+=('{"channel_threads":0, "feed_threads":0}')
+        ''
+        # configure different ports for the instances
+        + ''
+          configParts+=('{"port":${toString (cfg.port + scaleIndex)}}')
```
Not sure if we using nix strings would be a better idea here or even possible. Maybe something for another PR.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1318

RaitoBezarius commented:
@999eagle Can I ask you to resolve the comments and resolve the conflict? Then, we can go for the merge.

999eagle commented:
The conflict is in the release notes, so as long as the release notes for 24.05 are relatively short, it will conflict again very quickly. I've rebased to fix the conflict now.
I was still waiting on more feedback on some of the open comment threads but if you're fine with it I can just resolve them.

RaitoBezarius commented:
> The conflict is in the release notes, so as long as the release notes for 24.05 are relatively short, it will conflict again very quickly. I've rebased to fix the conflict now. I was still waiting on more feedback on some of the open comment threads but if you're fine with it I can just resolve them.

Don't hesitate to resolve them otherwise you get stuck in this "everyone is waiting on each other" hell and we can still improve further your stuff later on. :)
